### ISSUE TITLE
Add PKCE to user authentication

### ISSUE THREAD

**Issue by felix-hilden** [felix-hilden/tekore#189]
Spotify has announced support for the PKCE authentication method ([guide](https://developer.spotify.com/documentation/general/guides/authorization-guide/#authorization-code-flow-with-proof-key-for-code-exchange-pkce)). We should implement it. It seems that Spotify has documented it as optional, but reading the RFC, it is presented as an extension that should be used instead of the original one. So I think we won't provide any parameters to switch it off.

<strike>We should also consider the Implicit Grant flow and could it be useful. I think Spotify's disclaimer of it being a JS-only method put me off until now when I gave it more thought.</strike> It seems that the need for this does not exist, as it [cannot](https://stackoverflow.com/a/16747822/7089239) be used on a server, and PKCE is the preferred method for authentication that does not use a client secret.

Let's also add better documentation pulling all these authentication methods together.

**Comment by felix-hilden** [comment]
This raises a problem where each generated code challenge must be matched with a verifier it was generated with. I see two approaches.

- Store the previous verifier as class state. This has the obvious problem of not being able to handle parallel authentications.
- Send the verifier as `state`. It would need to be sent as plain text, making the whole enhancement pointless.
- Send a reference to a verifier as `state`. We could generate a sufficiently long random string to send as additional state. These strings would then map to verifiers and be deleted on use. In practice we would need to append state in case the user specifies it.

This needs to be thought out. Not having an ability to use whatever instance of `Credentials` for any request *could* be limiting. The thought of using some object to signify each authentication instance could also be useful. Let's see.

**Comment by felix-hilden** [comment]
Of course this can and should be implemented as an enhancement and deprecation of the original functionality, provided that the names don't clash.

**Comment by felix-hilden** [comment]
With the non-inclusion of Implicit Grant Flow, having one flow object could also be overkill. PKCE changes the refreshing of tokens too, so the refresh method needs to be changed anyway. Not using flow objects would allow for simpler changes.

```python
cred = tk.Credentials(client_id, client_secret, redirect_uri)
url, verifier = cred.user_authorisation_url(scope, state, verifier_bytes)
code, state_check = ...  # Redirect and get code and state
if state == state_check:
    token = cred.request_user_token(code, verifier)
else:
    raise SomeError('Important message.')
```

This wouldn't allow for automatically generating state, but a utility with good examples might be sufficient for that.

**Comment by felix-hilden** [comment]
Okay, so the Spotify authorisation guide states at the very end of its PKCE section, that its refresh token may be used only once to get a new access token, after which it is invalidated. So this is by no means a replacement to the regular user auth flow.

**Comment by felix-hilden** [comment]
Given no quick response from Spotify, we should implement PKCE as an option. So we ought to choose some names.
- 1/2 without PKCE: `user_authorisation_url`
- 2/2 without PKCE: `request_user_token`
- Refresh without PKCE: `refresh_user_token`
- 1/2 with PKCE: `pkce_user_authorisation`
- 2/2 with PKCE: `request_pkce_token`
- Refresh with PKCE: `refresh_pkce_token`

These are reasonably similar while also highlighting the fact that PKCE introduces special semantics.

**Sadly** this addition makes it impossible to distinguis between ordinary user refresh tokens and PKCE versions. This affects two things: `Credentials.refresh` and configuration. I'm leaning towards removing the type-agnostic refresh method. Self-refreshing tokens exists and it is easy to test for app vs user tokens. A note about that test should be added to the docs though. Configuration-wise, as the name of the refresh token variable can be configured, they can be saved and loaded even from the same file. But this kind of use shouldn't be expected, because a large part of the security is lost if ordinary tokens are also in use.

**Comment by felix-hilden** [comment]
And the plot thickens yet again. Seems like we could get away with a single refresh method for user tokens. But let's see how the situation develops [here](https://community.spotify.com/t5/Spotify-for-Developers/Client-ID-not-required-for-PKCE-refresh-despite-documentation/m-p/5025085).

**Comment by felix-hilden** [comment]
In the end, this doesn't require changes. We can naturally use the new boolean for tokens in deciding how to refresh them.

**Comment by felix-hilden** [comment]
Closed in #193
